dns/name.py

```diff
@@ -1082,3 +1121,178 @@ def from_wire(message: bytes, current: int) -> Tuple[Name, int]:
+    name = Name(new_labels)
+    if prefix_ok:
+        return _pad_to_max_name(name)
+    else:
+        return name
+
+
+def _absolute_successor(name: Name, origin: Name, prefix_ok: bool) -> Name:
+    # This is the RFC 4471 successor algorithm using the "absolute method" of section
+    # 3.1.2.
+    #
+    # Our caller must ensure that the name and origin are absolute, and that name is a
+    # subdomain of origin.
+    if prefix_ok:
+        # Try prefixing \000 as new label
+        try:
+            return _SUCCESSOR_PREFIX.concatenate(name)
+        except NameTooLong:
+            pass
+    while name != origin:
+        # Try extendinging the least significant label.
+        least_significant_label = name[0]
+        if len(least_significant_label) < 63:
+            # We may be able to extend the least label with a minimal additional byte.
+            # This is only "may" because we could have a maximal length name even though
+            # the least significant label isn't maximally long.
+            new_labels = [least_significant_label + _MINIMAL_OCTET]
+            new_labels.extend(name.labels[1:])
+            try:
+                return dns.name.Name(new_labels)
+            except dns.name.NameTooLong:
+                pass
+        # We can't extend the label either, so we'll try to increment the least
+        # signficant non-maximal byte in it.
+        octets = bytearray()
+        incremented = False
+        # We have to reverse here as we want to iterate from least-significant octet
+        # to most significant octet.
+        for octet in reversed(least_significant_label):
+            if not incremented:
+                if octet == _MAXIMAL_OCTET_VALUE:
+                    # We can't increment this, so drop it!
+                    continue
+                # Finally, something we can increment.  Note we have to
+                # apply a special rule for incremening "@", sending it to "[",
+                # because the standards say that when comparing names, uppercase
+                # letters compare as if they were their lower-case equivalents.
+                # If we increment "@" to "A", then it would compare as "a", which
+                # is after "[", "\", "]", "^", "_", and "`", so we would have
+                # skipped the most minimal successor, namely "[".
+                if octet == _AT_SIGN_VALUE:
+                    octet = _LEFT_SQUARE_BRACKET_VALUE
+                else:
+                    octet += 1
+                # Once we've incremented something, we just want to copy the rest,
+                # i.e. not do any further incrementing.
+                incremented = True
+            octets.append(octet)
+        if incremented:
+            # It worked!
+            #
+            # Our saved octets are backwards, so we have to reverse again.
+            new_labels = [bytes(reversed(octets))]
+            new_labels.extend(name.labels[1:])
```

Comment:
This seems like it could be simplified a lot.  Instead of iterating the reversed label and incrementally building a `bytearray`, and then reversing it, couldn't this just create the `bytearray` from the label at the start, and loop over a decreasing range?  It would like be faster, as well, as there wouldn't be reallocations, and the loop would terminate as soon as an increment was done (which would typically be on the first byte).

Reply:
All great points!